src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+                foreach (string item in newDesignTimeInputs)
+                {
+                    await FireTempPEDirtyAsync(item, value.ShouldCompile);
+                }
+            }
+            else
+            {
+                // Individual inputs dirty their PEs and possibly recompile
+                foreach (string item in addedDesignTimeInputs)
+                {
+                    await FireTempPEDirtyAsync(item, value.ShouldCompile);
+                }
+            }
+
+            // Shared items cause all TempPEs to be dirty, but don't recompile, to match legacy behaviour
+            if (hasRemovedDesignTimeSharedInputs || hasAddedDesignTimeSharedInputs)
+            {
+                // adding or removing shared design time inputs dirties things but doesn't recompile
+                foreach (string item in newDesignTimeInputs)
+                {
+                    // We don't want to fire again if we already fired above and compiled
+                    if (!addedDesignTimeInputs.Contains(item))
+                    {
+                        await FireTempPEDirtyAsync(item, false);
+                    }
+                }
+            }
+
+            foreach (string item in removedDesignTimeInputs)
+            {
+                _buildManager.Value.OnDesignTimeOutputDeleted(item);
+            }
+
+            bool TryGetValueIfUnused<T>(string item, ImmutableDictionary<string, T>.Builder source, ImmutableHashSet<string>.Builder otherSources, out T result)
+            {
+                // return the value from the source, but only if it doesn't appear in other sources
+                return source.TryGetValue(item, out result) && !otherSources.Contains(item);
+            }
+
+        }
+
+        protected virtual ITaskDelayScheduler CreateTaskScheduler()
+        {
+            return new TaskDelayScheduler(s_compilationWaitTime, UnconfiguredProjectServices.ThreadingService, DisposalToken);
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                DisposeTaskSchedulers();
+            }
+            await base.DisposeCoreAsync(initialized);
+        }
+
+        /// <summary>
+        /// InitializeInnerCoreAsync is responsible for setting an initial AppliedValue. This value will be used by any UI thread calls that may happen
+        /// before the first data flow blocks have been processed. If this method doesn't return a value then the system will block until the first blocks
+        /// have been applied but since we are initialized on the UI thread, that blocks for us, so we must return an initial state.
+        /// </summary>
+        protected override Task InitializeInnerCoreAsync(CancellationToken cancellationToken)
+        {
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem(), ImmutableDictionary.Create<NamedIdentity, IComparable>());
```

Comment:
Is this right? If I understand, by default it has zero monikers until you've processed your first changes? What if someone queries straight after load?

Reply:
Yes, that's correct. This is a rock-and-a-hard-place situation, where we can't initialize ProjectHostBridge on the UI thread (there was a thread on this with Lifeng in December) but we didn't want to initialize the TempPE manager unless it was needed by a designer.

Potentially we could change this now that we will have WinForms capabilities across project systems, which reduces the risk of unnecessary creation.

Reply:
How does someone right code that correctly queries this code at the right time?

Reply:
Discussed offline, hopefully will be able to move to a model where the initial value is correct without deadlocking.